File: framework/languages/java/samples/java/Bingo/Client/src/main/java/systems/zlink/samples/bingo/client/BingoNotificationInbox.java
package systems.zlink.samples.bingo.client;

import java.util.ArrayList;
import java.util.List;
import systems.zlink.samples.bingo.shared.contracts.Messages;

public final class BingoNotificationInbox {
    private final List<Messages.PlayerJoinedNotify> playerJoined = new ArrayList<>();
    private final List<Messages.BingoGameStartedNotify> started = new ArrayList<>();
    private final List<Messages.BingoNumberDrawnNotify> drawn = new ArrayList<>();
    private final List<Messages.BingoGameEndedNotify> ended = new ArrayList<>();

    public void playerJoined(Messages.PlayerJoinedNotify value) {
        playerJoined.add(value);
    }

    public void started(Messages.BingoGameStartedNotify value) {
        started.add(value);
    }

    public void drawn(Messages.BingoNumberDrawnNotify value) {
        drawn.add(value);
    }

    public void ended(Messages.BingoGameEndedNotify value) {
        ended.add(value);
    }

    public List<Messages.PlayerJoinedNotify> playerJoined() {
        return List.copyOf(playerJoined);
    }

    public List<Messages.BingoGameStartedNotify> started() {
        return List.copyOf(started);
    }

    public List<Messages.BingoNumberDrawnNotify> drawn() {
        return List.copyOf(drawn);
    }

    public List<Messages.BingoGameEndedNotify> ended() {
        return List.copyOf(ended);
    }

    public List<Messages.BingoWinner> winners() {
        if (ended.isEmpty()) {
            return List.of();
        }
        return ended.getLast().state().winners().stream()
            .map(Messages.BingoWinner::new)
            .toList();
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Client/src/main/kotlin/systems/zlink/samples/kotlin/bingo/client/BingoNotificationInbox.kt
package systems.zlink.samples.kotlin.bingo.client

import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameEndedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameStartedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoNumberDrawnNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoWinner
import systems.zlink.samples.kotlin.bingo.shared.contracts.PlayerJoinedNotify

class BingoNotificationInbox {
    private val playerJoinedValues = mutableListOf<PlayerJoinedNotify>()
    private val startedValues = mutableListOf<BingoGameStartedNotify>()
    private val drawnValues = mutableListOf<BingoNumberDrawnNotify>()
    private val endedValues = mutableListOf<BingoGameEndedNotify>()

    fun playerJoined(value: PlayerJoinedNotify) {
        playerJoinedValues += value
    }

    fun started(value: BingoGameStartedNotify) {
        startedValues += value
    }

    fun drawn(value: BingoNumberDrawnNotify) {
        drawnValues += value
    }

    fun ended(value: BingoGameEndedNotify) {
        endedValues += value
    }

    fun playerJoined(): List<PlayerJoinedNotify> = playerJoinedValues.toList()

    fun started(): List<BingoGameStartedNotify> = startedValues.toList()

    fun drawn(): List<BingoNumberDrawnNotify> = drawnValues.toList()

    fun ended(): List<BingoGameEndedNotify> = endedValues.toList()

    fun winners(): List<BingoWinner> =
        endedValues.lastOrNull()?.state?.winners?.map(::BingoWinner) ?: emptyList()
}
